### ISSUE TITLE
Concurrent bookings are (sometimes) impossible

### ISSUE THREAD

**Issue by JirkaChadima** [windingtree/wt-booking-api#8]
If we need to update on-chain data during a booking process (because we update availability), it is impossible to handle multiple bookings at once. This is basically the same problem we had with https://github.com/windingtree/wt-write-api/issues/42

**Comment by JirkaChadima** [comment]
Hi @umran, thanks for your comment. That is a viable solution if we were sure that there is only one existing queue that is issuing transactions on behalf of a single account (i. e. not only bookings, but also regular updates - such as rate plans, pictures etc.).

In an environment where we cannot guarantee that (cloud with multiple instances, write-api vs. booking-api) I think this might be problematic for concurrency reasons. Of course this can be mitigated by using a shared internal queue where all outgoing transactions for a single account are handled. But that's not something we currently have implemented.

Still, for rapidly changing data (which availability sure is), it's probably better to use a storage with mutable data on stable URI than rely on something like this. From what I've read there are many things that can [go wrong with nonce](https://github.com/windingtree/wt-write-api/issues/42#issuecomment-441058365).

**Comment by umran** [comment]
Hi @JirkaChadima, I guess I misunderstood the issue. I'm still a bit confused. The booking api currently uses an instance of adapter.js, which has a method called `updateAvailability`. If I understand it correctly, it is written so that all bookings issued by a single instance of the booking api is processed in order. This means that `updateAvailability` calls the write api in order as well. I'm not sure what goes on under the hood in the write api when it updates the onchain storage though, so I'm assuming the write api, in updating the onchain storage simply submits the transaction and returns without waiting for the transaction to be mined. If that is the case, then the issue can only be mitigated by modifying the write-api in some way.  But then again, that would only mitigate the issue on a per instance basis, because as you said, if multiple instances of the write-api is deployed, some external mechanism will have to be put in place to synchronize transactions across instances. Perhaps that adds more complexity to a potential client setup than is necessary.

Moving forward, do you expect this kind of conflict with concurrent bookings to be handled by client implementations? If so I think that is reasonable.

A potential client implementation could accept concurrent bookings by putting them in an internal queue that processes them one by one only after the most recent transaction is mined. Before any booking ends up in the queue, the client can verify it against local inventory, and on submission to the queue, it can update local inventory accordingly. Or I guess it could also circumvent the queue entirely and simply retry any failed transactions as per the error message.

**Comment by JirkaChadima** [comment]
> so I'm assuming the write api, in updating the onchain storage simply submits the transaction and returns without waiting for the transaction to be mined

It actually waits until the tx is mined. But there is no guarantee that only a single instance is running and that only a single instance of booking API is calling the write API.

All of your client-side proposals are probably possible. The optimal way (at least from my point of view), however, is to never touch the blockchain during the booking API call - i. e. update the availability in a storage that does not affect the document URIs and therefore does not affect the blockchain stored data.  

There are many options how and where this can be dealt with and I'm afraid that the best approach heavily depends on a particular setup of the Booking pipeline. This implementation is meant as a demo and very likely would break under the test of any real-life volume of traffic.

**Comment by umran** [comment]
I understand now what you mean by mutable storage under a stable URI. Yeah I think that would be a neat solution too.

**Comment by umran** [comment]
Hey @JirkaChadima I have also realized another issue if multiple booking api instances are in play at the same time. Race conditions. Right now for every booking, availability is read from the read-api and then updated via the write-api. If multiple booking-api instances issue concurrent independent bookings, there is a possibility of overbooking a particular room type. Perhaps you should consider making bookings transactional, and make the read and write operations involved in booking atomic.

**Comment by JirkaChadima** [comment]
Yeah, that's of course another problem. But again, this is just a sample implementation. Real-world thing would probably need to communicate with some existing system etc. I really thing this is beyond the scope of this implementation. But it's good to have these things in mind.

**Comment by JirkaChadima** [comment]
The Booking API now replies with 503 when the Write API is responding with 503 (it might be due to concurrency or other issues).

And the Booking API implementation is currently queueing the requests as well, but the clients might time out. So that's something to bear in mind as well.

I'm closing this issue, as I don't think we can do more to mitigate this with the current rather simplistic sample implementation.
